src/helpers/buffers/receive.rs

```diff
@@ -57,21 +92,37 @@ impl ReceiveBuffer {
     /// etc. It does not require all chunks to be of the same size, this assumption is baked in
     /// send buffers.
     pub fn receive_messages(&mut self, channel_id: &ChannelId, messages: &[u8]) {
+        // TODO: this is probably wrong. Need to maintain total count as
+        // unknown until we see a receive request.
+        //let total_records = channel_id.total_records.expect("can't receive without a known total record count");
+
         let offset = self
             .record_ids
             .entry(channel_id.clone())
             .or_insert_with(|| RecordId::from(0_u32));
+        *offset += 1;
+        // TODO: need to pre-increment and drop this reference so we can
+        // call update_record_count below.  Can we do better?
+        let offset = *offset;
 
         for msg in messages.chunks(MESSAGE_PAYLOAD_SIZE_BYTES) {
             let payload = msg.try_into().unwrap();
-            match self
+            let channel = self
```

Comment:
It does seem possible to move this construction out of the `for` loop and it could also solve the offset issue for you. I didn't think clearly when I wrote this, but now I think we can rely on `record_ids` and `inner` to be in sync and if there is nothing in `record_ids` for this channel, we can insert in both (`or_insert_with_key`?) and then use mutable references to receive channel and offset inside this for loop. 

maybe we can do even better and get rid of `record_ids` entirely, storing `(RecordId, ReceiveChannel)` pair inside `inner`